### PR TITLE
Transparent bridging for nullable values as Swift Optionals

### PR DESCRIPTION
**Pull request by chkn**
# Swift Optionals

Swift generally has no intrinsic support for arbitrary values being `null`. To express this, it provides an `Optional` type that you must use to wrap your value. In order to make this a natural experience for C# and F#, we don't expose the `Optional` type directly, but rather transparently bridge nullable values to that type. The following nullable types are supported:

- C# nullable reference types**
- `System.Nullable` (C# nullable value types)
- F# `Option` and `ValueOption` types

** C# nullable reference types are not reified in the type system--nullability is represented as attributes at the declaration site. Because of this, it's impossible for a generic type to determine if its type argument is a nullable reference type or not. This means that an accurate `SwiftType` can no longer be obtained from a static property on a generic type. Instead, I've created a new system for obtaining the `SwiftType`

# New SwiftType system

TL;DR: You should always use `SwiftType.Of` to obtain the `SwiftType` for any given managed type. There is a new optional argument where you can specify nullability. This argument is used when we reflect over the fields of a custom view to expose them to Swift. If we see a `NullableAttribute` on the field, we can now pass that information through to `SwiftType.Of` and the resulting `SwiftType` will then be bridged as an Optional value.

## Changes to Bindings

Instead of a public static property, we now use an attribute, `SwiftTypeAttribute`, to provide the `SwiftType`. The attribute has a virtual method `GetSwiftType` that is passed the `System.Type` on which the attribute is applied, as well as `SwiftType`s for any generic type arguments. By passing in the `SwiftType`s for the generic arguments, we can ensure that they take nullability into account.

In the general case, bindings can simply add a `SwiftImportAttribute` (which derives from `SwiftTypeAttribute`) with the library path from which the type is to be imported. That should be enough to handle generic and non-generic types. For opaque return types, or types with
a special mangling, you must specify the mangled name (for opaque return types, this is the mangled name for the descriptor).

This is a cleaner approach overall, and enables us to remove the hack for handling custom views from `SwiftType.Of`. Now, custom views are trivially handled by subclassing `SwiftTypeAttribute` to return a `CustomViewType`. Because of the way the attribute is inherited on subclasses, we can simply apply this internal subclass to our `View` base type, and then no attribute is needed by custom View subclasses implemented in user code.

### CODE REVIEW

File: src/SwiftUI/Mac/NSHostingView.cs [hunk 1]
```diff
@@ -12,13 +12,13 @@ public unsafe class NSHostingView : NSView, /*INSUserInterfaceValidations,*/ INS
 	{
 		public static NSHostingView Create (View view)
 		{
-			using (var handle = view.GetHandle ())
-				return Create (handle.Pointer, view.ViewType);
+			using (var handle = view.GetSwiftHandle ())
+				return Create (handle.Pointer, handle.SwiftType);
```
Comment (CartBlanche) [inline src/SwiftUI/Mac/NSHostingView.cs]: nit: In this case, could we not just pass `handle`, then obtain the Pointer/SwiftType inside the `Create` call?

File: XamMacSwiftUITest/AppDelegate.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ public class AppDelegate : NSApplicationDelegate
 		public override bool ApplicationShouldTerminateAfterLastWindowClosed (NSApplication sender)
 		{
 			GC.Collect ();
-			return true;
+			return false;
```
Comment (CartBlanche) [inline XamMacSwiftUITest/AppDelegate.cs:17]: Why don't we want the app to terminate properly? Same in the F# sample?

File: src/SwiftUI/SwiftUI/ModifiedOpacity.cs [hunk 3]
```diff
@@ -4,18 +4,11 @@
 
 namespace SwiftUI
 {
+    [SwiftImport (SwiftUILib.Path, "$s7SwiftUI4ViewPAAE7opacityyQrSdFQOMQ")]
```
Comment (CartBlanche) [inline src/SwiftUI/SwiftUI/ModifiedOpacity.cs:7]: Nice!

**Comment by CartBlanche** [comment]
I'm really like this refactor. Removes a fair amount of duplication and I think overall simplifies things.

File: XamMacSwiftUITest/AppDelegate.cs (hunk 2 above)
Comment (chkn) [reply XamMacSwiftUITest/AppDelegate.cs:17]: The app is crashing when you close the window, meaning we have another memory management issue. Having it terminate serves no purpose and hides these types of issues

File: src/SwiftUI/Mac/NSHostingView.cs (hunk 1 above)
Comment (chkn) [reply src/SwiftUI/Mac/NSHostingView.cs]: Great idea! I also took the opportunity to switch to using regular constructors instead of `Create` static methods

File: XamMacSwiftUITest/AppDelegate.cs (hunk 2 above)
Comment (CartBlanche) [reply XamMacSwiftUITest/AppDelegate.cs:17]: Oh another leak!
